rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/adapter/EitherTypeAdapter.java

```diff
@@ -0,0 +1,138 @@
+
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.gson.TypeAdapter;
+import com.google.gson.TypeAdapterFactory;
+import com.google.gson.reflect.TypeToken;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import java.io.IOException;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.function.Predicate;
+import javax.annotation.Nullable;
+import org.sonarsource.sonarlint.core.rpc.protocol.Either;
+
+import static org.sonarsource.sonarlint.core.rpc.protocol.Lsp4jUtils.isEither;
+
+public class EitherTypeAdapter<L, R> extends TypeAdapter<Either<L, R>> {
+
+  private final org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapter<L, R> lsp4jEitherAdapter;
+
+  public EitherTypeAdapter(Gson gson, TypeToken<? extends Either<L, R>> typeToken, @Nullable Predicate<JsonElement> leftChecker,
+    @Nullable Predicate<JsonElement> rightChecker) {
+    var eitherType = (ParameterizedType) typeToken.getType();
+    var lsp4jEitherType = new ParameterizedTypeImpl(org.eclipse.lsp4j.jsonrpc.messages.Either.class,
+      org.eclipse.lsp4j.jsonrpc.messages.Either.class, eitherType.getActualTypeArguments());
+    var lsp4jTypeToken = (TypeToken<? extends org.eclipse.lsp4j.jsonrpc.messages.Either<L, R>>) TypeToken.get(lsp4jEitherType);
+    this.lsp4jEitherAdapter = new org.eclipse.lsp4j.jsonrpc.json.adapters.EitherTypeAdapter<>(gson, lsp4jTypeToken, leftChecker,
+      rightChecker, null, null);
+  }
+
+  public EitherTypeAdapter(Gson gson, TypeToken<? extends Either<L, R>> typeToken) {
+    this(gson, typeToken, null, null);
+  }
+
+  public static class Factory implements TypeAdapterFactory {
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    @Override
+    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> typeToken) {
+      if (!isEither(typeToken.getType())) {
+        return null;
+      }
+      return new EitherTypeAdapter(gson, typeToken);
+    }
+  }
+
+  @Override
+  public void write(JsonWriter out, Either<L, R> value) throws IOException {
+    this.lsp4jEitherAdapter.write(out, value.getLsp4jEither());
+  }
+
+  @Override
+  public Either<L, R> read(JsonReader in) throws IOException {
+    return new Either<>(this.lsp4jEitherAdapter.read(in));
+  }
+
+  private static class ParameterizedTypeImpl implements ParameterizedType {
+
+    private final Type ownerType;
+    private final Type rawType;
+    private final Type[] actualTypeArguments;
+
+    ParameterizedTypeImpl(ParameterizedType original, Type[] typeArguments) {
+      this(original.getOwnerType(), original.getRawType(), typeArguments);
+    }
+
+    ParameterizedTypeImpl(Type ownerType, Type rawType, Type[] typeArguments) {
+      this.ownerType = ownerType;
+      this.rawType = rawType;
+      this.actualTypeArguments = typeArguments;
+    }
+
+    @Override
+    public Type getOwnerType() {
+      return ownerType;
```

Comment:
After reading the Javadoc for this class, I think we could return null here everytime